dbms/src/Flash/Mpp/MPPTunnelSet.cpp

```diff
@@ -63,83 +63,200 @@ void MPPTunnelSetBase<Tunnel>::write(tipb::SelectResponse & response)
+            {
+                local_cnt++;
+                if (local_cnt == local_tunnel_cnt)
+                    tunnels[i]->write(std::move(ori_tracked_packet));
+                else
+                    tunnels[i]->write(ori_tracked_packet->copy()); // NOLINT
+            }
+            else
+            {
+                remote_cnt++;
+                if (remote_cnt == remote_tunnel_cnt)
+                    tunnels[i]->write(std::move(remote_tracked_packet));
+                else
+                    tunnels[i]->write(remote_tracked_packet->copy()); // NOLINT
+            }
         }
-        break;
     }
-    case CompressionMethod::LZ4:
+
+    if (is_broadcast)
     {
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_lz4_compression).Increment(sz);
-        break;
+        UPDATE_EXCHANGE_MATRIC(broadcast, CompressionMethod::NONE, local_tunnel_cnt * ori_tracked_packet_bytes, local_tunnel_cnt * ori_tracked_packet_bytes, true);
+        UPDATE_EXCHANGE_MATRIC(broadcast, compression_method, remote_tunnel_cnt * ori_tracked_packet_bytes, remote_tunnel_cnt * remote_tracked_packet_bytes, false);
     }
-    case CompressionMethod::ZSTD:
+    else
     {
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_zstd_compression).Increment(sz);
-        break;
-    }
-    default:
-        break;
+        UPDATE_EXCHANGE_MATRIC(passthrough, CompressionMethod::NONE, local_tunnel_cnt * ori_tracked_packet_bytes, local_tunnel_cnt * ori_tracked_packet_bytes, true);
+        UPDATE_EXCHANGE_MATRIC(passthrough, compression_method, remote_tunnel_cnt * ori_tracked_packet_bytes, remote_tunnel_cnt * remote_tracked_packet_bytes, false);
     }
 }
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks)
+void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks, bool is_broadcast)
 {
     RUNTIME_CHECK(!tunnels.empty());
     auto && tracked_packet = MPPTunnelSetHelper::ToPacketV0(blocks, result_field_types);
     if (!tracked_packet)
         return;
+    const size_t local_tunnel_cnt = std::accumulate(tunnels.begin(), tunnels.end(), 0, [](auto res, auto && tunnel) {
+        return res + tunnel->isLocal();
+    });
+    size_t tracked_packet_bytes = tracked_packet->getPacket().ByteSizeLong();
+    BroadcastOrPassThroughWriteImpl(
+        tunnels,
+        local_tunnel_cnt,
+        is_broadcast,
+        std::move(tracked_packet),
+        tracked_packet_bytes,
+        nullptr,
+        CompressionMethod::NONE);
+}
+
+template <typename Tunnel>
+void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks, MPPDataPacketVersion version, CompressionMethod compression_method, bool is_broadcast)
+{
+    if (MPPDataPacketV0 == version)
+        return broadcastOrPassThroughWrite(blocks, is_broadcast);
+
+    RUNTIME_CHECK(!tunnels.empty());
+    const size_t local_tunnel_cnt = std::accumulate(tunnels.begin(), tunnels.end(), 0, [](auto res, auto && tunnel) {
```

Comment:
extract local tunnel info to field member of MPPTunnelSet at initial time?

Reply:
`tunnels` can be changed by `MPPTunnelSetBase<Tunnel>::registerTunnel`.

Reply:
Yeah, in my understanding, these info will be settle down in preparation time and won't change any more, thus no need to calculate each time.

Reply:
Fixed. Use another member variables to present it.